### PR TITLE
Mongodb Logging Receiver

### PR DESCRIPTION
**Pull request by schmikei**
Adds Initial File Based Logging Receiver Solution for mongodb

Design as of submission of this PR has been spec-ed but not reviewed.

Tested on mongodb community versions:
- 5.0
- 4.4
- 4.2
- 3.0

It is important to know that at Mongo 4.4, the application changed log formats from single lines to JSON objects. This attempt aims to write a FluentBit config that requires no user input to send logs to Google. Because of this, we needed two parsers for the same log type (they aren't different logs but differently formatted). 

Config is relatively simplistic:
```yaml
 include_paths: ['/var/log/mongodb/mongod.log*]  # files to include in the tail including log rotations
 exclude_paths: [] # files to exclude tailing if desired
```

Important Notes:
- WiredTiger messages are promoted to the message field
- Severity is properly parsed but is persisted in order to keep Debug context information on the log
- Renames a couple fields to be consistent across all mongo versions

Sample Input (JSON): 
```JSON
{
  "t": {
    "$date": "2021-12-09T15:30:48.514-05:00"
  },
  "s": "I",
  "c": "STORAGE",
  "id": 22430,
  "ctx": "Checkpointer",
  "msg": "WiredTiger message",
  "attr": {
    "message": "[1638908471:104686][3028:0x7f3dd58b6700], WT_SESSION.checkpoint: [WT_VERB_CHECKPOINT_PROGRESS] saving checkpoint snapshot min: 1522, snapshot max: 1522 snapshot count: 0, oldest timestamp: (0, 0) , meta checkpoint timestamp: (0, 0) base write gen: 5587"
  }
}
```

Parses into (roughly):
```json
{
  "insertId": "1kw9mkvg5f4cr2u",
  "jsonPayload": {
    "component": "STORAGE",
    "context": "Checkpointer",
    "severity": "I",
    "id": 22430,
    "message": "[1639066405:805863][2411:0x7f3dd270b700], WT_SESSION.checkpoint: [WT_VERB_CHECKPOINT_PROGRESS] saving checkpoint snapshot min: 5441, snapshot max: 5441 snapshot count: 0, oldest timestamp: (0, 0) , meta checkpoint timestamp: (0, 0) base write gen: 1"
  },
  "resource": {
    "type": "gce_instance",
    "labels": {
      "zone": "us-central1-b",
      "project_id": "otel-agent-dev",
      "instance_id": "5455835670130069851"
    }
  },
  "timestamp": "2021-12-09T16:15:30.485146325Z",
  "severity": "INFO",
  "logName": "projects/otel-agent-dev/logs/mongodb_default",
  "receiveTimestamp": "2021-12-09T16:15:31.654155897Z"
}
```

Sample input (Regex):
```sh
2021-12-09T19:30:28.448+0000 I  NETWORK  [conn4] received client metadata from 127.0.0.1:47320 conn4: { application: { name: "MongoDB Shell" }, driver: { name: "MongoDB Internal Client", version: "4.2.15" }, os: { type: "Linux", name: "Ubuntu", architecture: "x86_64", version: "20.04" } }
```

Parses into (roughly):
``` json
{
  "insertId": "12uotejg4scg1wv",
  "jsonPayload": {
    "severity": "I",
    "component": "NETWORK",
    "message": "received client metadata from 127.0.0.1:47320 conn4: { application: { name: \"MongoDB Shell\" }, driver: { name: \"MongoDB Internal Client\", version: \"4.2.15\" }, os: { type: \"Linux\", name: \"Ubuntu\", architecture: \"x86_64\", version: \"20.04\" } }",
    "context": "conn4"
  },
  "resource": {
    "type": "gce_instance",
    "labels": {
      "zone": "us-central1-b",
      "project_id": "otel-agent-dev",
      "instance_id": "3982970403694615357"
    }
  },
  "timestamp": "2021-12-09T19:30:28.448Z",
  "severity": "INFO",
  "logName": "projects/otel-agent-dev/logs/mongodb_default",
  "receiveTimestamp": "2021-12-09T19:30:29.365202511Z"
}
```
This is my first attempt at contribution towards this project so feel free to clarify conventions that were not followed or if any refactors you would like to see in a separate PR.

### CODE REVIEW

File: apps/mongodb.go
Comment (qingling128) [inline apps/mongodb.go:112]: We tend to remove the `severity` field from `jsonPayload` after extracting the value and moving it to the top level `severity` field. Is that something we could do for MongoDB as well?

File: confgenerator/testdata/valid/linux/logging-receiver_mongodb/golden_fluent_bit_main.conf
Comment (qingling128) [inline confgenerator/testdata/valid/linux/logging-receiver_mongodb/golden_fluent_bit_main.conf:275]: The filters from line 65 - 275 seem pretty expensive.

Can we use the built-in feature like severity_key instead? https://github.com/fluent/fluent-bit/blob/0ba2dc2196e4c2c19a6075f8154afe7be137119d/plugins/out_stackdriver/stackdriver_conf.c#L293

File: apps/mongodb.go
Comment (qingling128) [inline apps/mongodb.go:1]: Can we add a MD file (aka documentation) as well? https://github.com/GoogleCloudPlatform/ops-agent/tree/master/docs

File: apps/mongodb.go
Comment (qingling128) [inline apps/mongodb.go:129]: [Optional] We could compress this section to save some space and improve readability. Similar to https://github.com/GoogleCloudPlatform/ops-agent/blob/d88a768910a7f4eb2a4bbae94bb829ea9effa261/apps/nginx.go#L146-L158

File: apps/mongodb.go
Comment (qingling128) [inline apps/mongodb.go:189]: Same here. We could compress them like https://github.com/GoogleCloudPlatform/ops-agent/blob/d88a768910a7f4eb2a4bbae94bb829ea9effa261/apps/nginx.go#L146-L158 for space saving and readability.

File: confgenerator/testdata/valid/linux/logging-receiver_mongodb/golden_fluent_bit_main.conf
Comment (qingling128) [reply confgenerator/testdata/valid/linux/logging-receiver_mongodb/golden_fluent_bit_main.conf:275]: In fact, that might having this app use its own output plugin instance, which comes with some concerns as well.

I'm trying to think what other ways to reduce the complexity, when I noticed that

https://github.com/GoogleCloudPlatform/ops-agent/pull/287/files#diff-56a349cfc933ac75b1c65de7a835cfce442abcb88d21628c7c4db0c240620a61R119-R177 and https://github.com/GoogleCloudPlatform/ops-agent/pull/287/files#diff-56a349cfc933ac75b1c65de7a835cfce442abcb88d21628c7c4db0c240620a61R210-R268 seem identical? Do we need both?

Comment (schmikei) [reply confgenerator/testdata/valid/linux/logging-receiver_mongodb/golden_fluent_bit_main.conf:275]: As far as the severity key link you posted, it is looking to process the stackdriver output of the fluent bit entry right? As in it is looking for a "severity_key" on the entry or the [default](https://github.com/fluent/fluent-bit/blob/0ba2dc2196e4c2c19a6075f8154afe7be137119d/plugins/out_stackdriver/stackdriver.h#L55).

What these lines are doing is mapping severity to their respective https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry#logseverity GCL severity. 

As for the second comment about the similar blocks, they are essentially doing two of the same things although we need both of them because we have two parsers being run because we try to auto-detect between json and regex based logs which in fluent bit seems to take the approach of just add a separate parser for each.

File: apps/mongodb.go
Comment (schmikei) [reply apps/mongodb.go:189]: These are for a single field rename though kind of like a mapping right? Like for that one it only looking at the source key `level` and the moving to `logging.googleapis.com/severity`. I guess I could create some more analogous helper function but opted not to use the `TranslationComponents` because these are renaming multiple fields to be consistent. 

I also didn't want to break that function by chance since a lot of integrations currently rely on it, but if you can think of a way I would be interested!

**Review by schmikei — COMMENTED** [review]
Sorry just getting around to responding to this, but thanks for the feedback I've got some small stuff to look into as far as the JSON parser bit and to some weird assumptions it makes but I will do that sometime today.

I did have some questions on some of the suggestions if you don't mind taking a look.

Thanks again @qingling128

File: apps/mongodb.go
Comment (schmikei) [reply apps/mongodb.go:129]: Ahh I misunderstood when i first read, you just want to omit the extra whitespace/unneeeded declarations. Sorry about that!

File: confgenerator/testdata/valid/linux/logging-receiver_mongodb/golden_fluent_bit_main.conf
Comment (schmikei) [reply confgenerator/testdata/valid/linux/logging-receiver_mongodb/golden_fluent_bit_main.conf:205]: You ended up being right these were getting duplicated on both parsers. After re-examining I shared the severity parser portion on both the JSON and Regex based logs 👍

**Comment by qingling128** [comment]
Is this one still WIP or waiting for another review?

**Comment by schmikei** [comment]
@qingling128 still need to add documentation, will ping you once that is finished!

**Comment by schmikei** [comment]
@qingling128 I believe I have resolved the feedback. Is there a way you could take another look? Thanks in advance!

File: apps/mongodb.go
Comment (qingling128) [reply apps/mongodb.go:1]: Thanks! Looks like the tests are not executed though: This file is missing Mongodb? https://github.com/GoogleCloudPlatform/ops-agent/blob/051df889b02eba0d08e0087dd1ebfc8bb05acf25/integration_test/third_party_apps_data/agent/ops-agent/linux/supported_applications.txt

File: confgenerator/testdata/valid/linux/logging-receiver_mongodb/golden_fluent_bit_main.conf
Comment (qingling128) [reply confgenerator/testdata/valid/linux/logging-receiver_mongodb/golden_fluent_bit_main.conf:207]: Great, thanks!

File: confgenerator/fluentbit/modify.go
Comment (qingling128) [inline confgenerator/fluentbit/modify.go:1]: @quentinmit - Since you are working on a generic`modify_fields` processor, could you help take a look at this file to see if there are any concerns?

File: confgenerator/fluentbit/modify.go
Comment (quentinmit) [inline confgenerator/fluentbit/modify.go:7]: `= string` causes these to be type aliases instead of new types; you probably meant to say `type ModifyRule string`.

File: confgenerator/fluentbit/modify.go
Comment (quentinmit) [reply confgenerator/fluentbit/modify.go:1]: These look like helpers for generating fluentbit config directly. I'm not particularly worried about what that API looks like - whatever makes the processors simple and readable is good. I doubt we'll use any of this code for the generic processor implementation, though.

**Review by qingling128 — APPROVED** [review]
LGTM

File: integration_test/third_party_apps_data/agent/ops-agent/linux/enable_mongodb
Comment (qingling128) [reply integration_test/third_party_apps_data/agent/ops-agent/linux/enable_mongodb]: Ignore this for now.

Comment (qingling128) [reply integration_test/third_party_apps_data/agent/ops-agent/linux/enable_mongodb]: Ah, looks like this has been applied and caused test failures. I did not mean to send this one as I was thinking about a follow up PR for the overall cleanup of everything that includes `{APP}_default`. We should try avoid `_default` in log receiver IDs as the log name that is displayed in the Log Viewer is derived from it.

Instead, in the sample configurations, we should just use the receiver type name as the sample receiver ID. This way the default log queries are unified.

Comment (quentinmit) [reply integration_test/third_party_apps_data/agent/ops-agent/linux/enable_mongodb]: I agree with the suggestion to use the receiver type name as the receiver and pipeline names in the sample config.

**Comment by schmikei** [comment]
Integration test seems to be memcached on sles 12 don't think related. I can probably rerun if desired! 
image

**Comment by qingling128** [comment]
Needs rebasing before it can be merged.

**Comment by schmikei** [comment]
image

**Comment by schmikei** [comment]
@qingling128 I believe the third party test failure is unrelated to these changes.

File: apps/modify/modify.go
Comment (quentinmit) [inline apps/modify/modify.go:1]: This package needs a package comment (and can we put it under `.../confgenerator/fluentbit/modify` since it appears to be helpers for generating fluentbit config?)

File: apps/mongodb.go
Comment (quentinmit) [inline apps/mongodb.go:45]: Can you add a link to the mongodb docs for the json log format to this comment?

File: integration_test/third_party_apps_data/applications/mongodb/debian_ubuntu/install
Comment (quentinmit) [inline integration_test/third_party_apps_data/applications/mongodb/debian_ubuntu/install:34]: Is there anything we can do to test the other supported versions of mongodb?

File: integration_test/third_party_apps_data/applications/mongodb/debian_ubuntu/install
Comment (quentinmit) [inline integration_test/third_party_apps_data/applications/mongodb/debian_ubuntu/install:9]: You can use `$(lsb_release -sc)` to get the codename and switch on `$(lsb_release -si)` which is `Ubuntu` or `Debian` to simplify the case and futureproof, though I guess you still want to hardcode Debian 11 for now since it needs buster packages.

File: integration_test/third_party_apps_data/applications/mongodb/debian_ubuntu/post
Comment (quentinmit) [inline integration_test/third_party_apps_data/applications/mongodb/debian_ubuntu/post:3]: Does mongod need to be restarted here?

File: integration_test/third_party_apps_data/applications/mongodb/debian_ubuntu/install
Comment (quentinmit) [inline integration_test/third_party_apps_data/applications/mongodb/debian_ubuntu/install:42]: Does mongod need to be enabled here?

**Review by quentinmit — APPROVED** [review]
I left some comments, just some style/simplification nits. I'll mark approved so you can merge after those are addressed.

File: integration_test/third_party_apps_data/applications/mongodb/debian_ubuntu/post
Comment (schmikei) [reply integration_test/third_party_apps_data/applications/mongodb/debian_ubuntu/post:3]: Mongo by default install does not start I believe.

Spiked out a fresh run on a new VM and this line at least starts the dead/inactive service.
```
Created symlink /etc/systemd/system/multi-user.target.wants/mongod.service → /lib/systemd/system/mongod.service.
keith_schmitt_bluemedora_com@keith-testing-mongo-debian-11:~$ mongosh
Current Mongosh Log ID:	61fbef1c2f64cf0750e81915
Connecting to:		mongodb://127.0.0.1:27017/?directConnection=true&serverSelectionTimeoutMS=2000&appName=mongosh+1.1.9
MongoNetworkError: connect ECONNREFUSED 127.0.0.1:27017
keith_schmitt_bluemedora_com@keith-testing-mongo-debian-11:~$ mongo
MongoDB shell version v5.0.5
connecting to: mongodb://127.0.0.1:27017/?compressors=disabled&gssapiServiceName=mongodb
Error: couldn't connect to server 127.0.0.1:27017, connection attempt failed: SocketException: Error connecting to 127.0.0.1:27017 :: caused by :: Connection refused :
connect@src/mongo/shell/mongo.js:372:17
@(connect):2:6
exception: connect failed
exiting with code 1
keith_schmitt_bluemedora_com@keith-testing-mongo-debian-11:~$ sudo systemctl status mongod
● mongod.service - MongoDB Database Server
     Loaded: loaded (/lib/systemd/system/mongod.service; enabled; vendor preset: enabled)
     Active: inactive (dead)
       Docs: https://docs.mongodb.org/manual
```

File: integration_test/third_party_apps_data/applications/mongodb/debian_ubuntu/install
Comment (schmikei) [reply integration_test/third_party_apps_data/applications/mongodb/debian_ubuntu/install:42]: I do not think so, I'll remove it!

File: integration_test/third_party_apps_data/applications/mongodb/debian_ubuntu/install
Comment (schmikei) [reply integration_test/third_party_apps_data/applications/mongodb/debian_ubuntu/install:28]: Hmm we could try to specify the default version of mongodb per os-release... however I've been burnt by not pinning versions. We could arbitrarily install versions on certain releases of debian/ubuntu but then the test becomes less focused and harder to debug potentially. Perhaps something we can solve later? I'm open to opinions or further enhancements. Have any thoughts @quentinmit @qingling128 ?

Comment (qingling128) [reply integration_test/third_party_apps_data/applications/mongodb/debian_ubuntu/install:28]: Yeah, I think this is OK for the first iteration.

Separately, we should have some discussion about testing different versions of the application. But better punt that until the effort to reduce flakiness lands (aka reduce the scope of tests that are triggered based on what files the PR touches). With the current setup, if we try to test all supported app version on all distros for all PRs, the flakiness and quota issues are gonna bite us hard. It also feels like an overkill to test all the combinations.

CC @martijnvans for visibility in case anything is not accurate.

File: integration_test/third_party_apps_data/applications/mongodb/debian_ubuntu/post
Comment (quentinmit) [reply integration_test/third_party_apps_data/applications/mongodb/debian_ubuntu/post:3]: Huh, okay. That's weird on the part of the mongodb package - Debian policy says that daemons should be enabled and started automatically by package installation unless they require user configuration before they can run (and you don't appear to have that here). I guess that's the danger of using non-distribution packages. You could simplify the install script to `systemctl enable --now mongod` to combine enable and start.

File: integration_test/third_party_apps_data/applications/mongodb/debian_ubuntu/install
Comment (quentinmit) [reply integration_test/third_party_apps_data/applications/mongodb/debian_ubuntu/install:28]: +1, fine to punt for this PR and think about for the future.

**Review by quentinmit — APPROVED** [review]

**Review by qingling128 — APPROVED** [review]
LGTM. Triggering integration tests for a final check before merging.

**Comment by schmikei** [comment]
```
--- FAIL: TestThirdPartyApps (0.00s)
    --- PASS: TestThirdPartyApps/centos-7 (0.00s)
        --- SKIP: TestThirdPartyApps/centos-7/apache (0.00s)
        --- SKIP: TestThirdPartyApps/centos-7/redis (0.00s)
        --- SKIP: TestThirdPartyApps/centos-7/cassandra (0.00s)
        --- SKIP: TestThirdPartyApps/centos-7/nginx (0.02s)
        --- SKIP: TestThirdPartyApps/centos-7/mysql (0.00s)
        --- SKIP: TestThirdPartyApps/centos-7/jvm (0.00s)
        --- PASS: TestThirdPartyApps/centos-7/memcached (128.94s)
        --- PASS: TestThirdPartyApps/centos-7/mongodb (137.70s)
        --- PASS: TestThirdPartyApps/centos-7/postgresql (147.63s)
        --- PASS: TestThirdPartyApps/centos-7/tomcat (165.60s)
        --- PASS: TestThirdPartyApps/centos-7/elasticsearch (190.40s)
    --- PASS: TestThirdPartyApps/ubuntu-2004-lts (0.00s)
        --- SKIP: TestThirdPartyApps/ubuntu-2004-lts/apache (0.00s)
        --- SKIP: TestThirdPartyApps/ubuntu-2004-lts/mysql (0.00s)
        --- SKIP: TestThirdPartyApps/ubuntu-2004-lts/nginx (0.00s)
        --- SKIP: TestThirdPartyApps/ubuntu-2004-lts/redis (0.00s)
        --- SKIP: TestThirdPartyApps/ubuntu-2004-lts/jvm (0.00s)
        --- SKIP: TestThirdPartyApps/ubuntu-2004-lts/cassandra (0.00s)
        --- PASS: TestThirdPartyApps/ubuntu-2004-lts/tomcat (181.91s)
        --- PASS: TestThirdPartyApps/ubuntu-2004-lts/mongodb (196.55s)
        --- PASS: TestThirdPartyApps/ubuntu-2004-lts/memcached (201.33s)
        --- PASS: TestThirdPartyApps/ubuntu-2004-lts/elasticsearch (247.76s)
        --- PASS: TestThirdPartyApps/ubuntu-2004-lts/postgresql (285.86s)
    --- FAIL: TestThirdPartyApps/debian-10 (0.00s)
        --- SKIP: TestThirdPartyApps/debian-10/mysql (0.00s)
        --- FAIL: TestThirdPartyApps/debian-10/postgresql (6.38s)
        --- PASS: TestThirdPartyApps/debian-10/redis (124.65s)
        --- PASS: TestThirdPartyApps/debian-10/mongodb (127.64s)
        --- PASS: TestThirdPartyApps/debian-10/memcached (129.90s)
        --- PASS: TestThirdPartyApps/debian-10/apache (139.53s)
        --- PASS: TestThirdPartyApps/debian-10/jvm (143.68s)
        --- PASS: TestThirdPartyApps/debian-10/nginx (152.69s)
        --- PASS: TestThirdPartyApps/debian-10/cassandra (165.10s)
        --- PASS: TestThirdPartyApps/debian-10/elasticsearch (177.12s)
        --- PASS: TestThirdPartyApps/debian-10/tomcat (311.98s)
    --- PASS: TestThirdPartyApps/sles-12 (0.00s)
        --- SKIP: TestThirdPartyApps/sles-12/apache (0.00s)
        --- SKIP: TestThirdPartyApps/sles-12/mysql (0.00s)
        --- SKIP: TestThirdPartyApps/sles-12/redis (0.00s)
        --- SKIP: TestThirdPartyApps/sles-12/cassandra (0.00s)
        --- SKIP: TestThirdPartyApps/sles-12/jvm (0.00s)
        --- SKIP: TestThirdPartyApps/sles-12/nginx (0.00s)
        --- PASS: TestThirdPartyApps/sles-12/tomcat (467.99s)
        --- PASS: TestThirdPartyApps/sles-12/elasticsearch (477.18s)
        --- PASS: TestThirdPartyApps/sles-12/mongodb (610.64s)
        --- PASS: TestThirdPartyApps/sles-12/memcached (655.89s)
        --- PASS: TestThirdPartyApps/sles-12/postgresql (688.34s)
    --- PASS: TestThirdPartyApps/sles-15 (0.00s)
        --- SKIP: TestThirdPartyApps/sles-15/apache (0.00s)
        --- SKIP: TestThirdPartyApps/sles-15/cassandra (0.00s)
        --- SKIP: TestThirdPartyApps/sles-15/nginx (0.00s)
        --- SKIP: TestThirdPartyApps/sles-15/mysql (0.00s)
        --- SKIP: TestThirdPartyApps/sles-15/redis (0.00s)
        --- SKIP: TestThirdPartyApps/sles-15/jvm (0.00s)
        --- PASS: TestThirdPartyApps/sles-15/mongodb (500.37s)
        --- PASS: TestThirdPartyApps/sles-15/memcached (505.68s)
        --- PASS: TestThirdPartyApps/sles-15/tomcat (697.59s)
        --- PASS: TestThirdPartyApps/sles-15/elasticsearch (708.93s)
        --- PASS: TestThirdPartyApps/sles-15/postgresql (889.24s)
FAIL
FAIL	command-line-arguments	889.327s
FAIL
```

Seems like a flaky postgres? Mongo tests passed fine

**Pull request merged**
